src/v/cluster/topic_recovery_validator.cc

```diff
@@ -0,0 +1,376 @@
+// remote.manifest_exists(bucket, path,
+// rtc).handle_exception_type([](terminate_early_exception) { return
+// early_termination; });` on the caller side, this local abort_source will
+// propagate root abort requests and local terminate_early requests
+// `ss::abort_source validation_as{}; auto master_as_subscription =
+// as.subscribe([&] { validation_as.request_abort(); });` ultimately this design
+// is not implemented here, since there seems no concrete use case to terminate
+// early at this level. either it's a general shutdown case, or the caller code
+// can get a better sense of what's going on by knowning the result of all
+// partitions
+
+namespace cluster {
+
+auto do_validate_manifest_existence(
+  cloud_storage_clients::bucket_name const& bucket,
+  model::ntp const& ntp,
+  model::initial_revision_id rev_id,
+  cloud_storage::remote& remote,
+  retry_chain_node& rtc) -> ss::future<validation_result> {
+    // check manifest existence, handle download_result::success as Found
+    // and the rest as NotFound (but warn for failed and timedout).
+    // This existence check could be flattened in anomalies_detector, but
+    // it's not a perfect overlap with the functionalities, so it's
+    // performed externally
+    using enum validation_result;
+
+    return remote.partition_manifest_exists(bucket, ntp, rev_id, rtc)
+      .then(
+        [ntp, rev_id](cloud_storage::remote::partition_manifest_existence res) {
+            auto [download_res, _] = res;
+
+            // only res==success (manifest found) or res==notfound (manifest NOT
+            // found) make sense. warn for timedout and failed
+            if (download_res == cloud_storage::download_result::success) {
+                vlog(clusterlog.info, "{}/{} manifest found", ntp, rev_id);
+                return passed;
+            }
+
+            if (download_res == cloud_storage::download_result::notfound) {
+                vlog(clusterlog.info, "{}/{} manifest NOT found", ntp, rev_id);
+                return missing_manifest;
+            }
+
+            // abnormal failure mode: could be a configuration issue or an
+            // external service issue (note that the manifest path will end in
+            // .bin but the value is just a hint of the HEAD request that
+            // generated the abnormal result)
+            vlog(
+              clusterlog.error,
+              "{}/{} ({}) manifest NOT found. download_result: {}",
+              ntp,
+              rev_id,
+              cloud_storage::generate_partition_manifest_path(ntp, rev_id, {}),
+              download_res);
+
+            return download_issue;
+        });
+}
+
+auto do_validate_manifest_metadata(
+  cloud_storage_clients::bucket_name const& bucket,
+  model::ntp const& ntp,
+  model::initial_revision_id rev_id,
+  cloud_storage::remote& remote,
+  retry_chain_node& rtc,
+  recovery_checks checks) -> ss::future<validation_result> {
+    using enum validation_result;
```

Comment:
This is cool in a smaller scope, but for larger functions like this I think it's valuable to keep the explicit enums stated. For instance, it isn't immediately clear if `failure` is referring to a member variable that I missed, or an enum